### PR TITLE
Incorporate pisa - part 3

### PR DESCRIPTION
**Pull request by apepper**

### CODE REVIEW

File: src/Data/DataIndexResponse.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { RawItem } from './RawItem'
+
+export interface DataIndexResponse {
+  results: RawItem[]
+  count?: number
+  continuation?: string
+}
```
Comment (dcsaszar) [inline src/Data/DataIndexResponse.ts:7]: * [ ] FYI As long as there are no plans to add non-type exports to things like `DataIndexResponse` and `RawItem` a `Data/types.ts` could make more sense.

File: src/Data/toClientParams.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    _order: params.order().length
+      ? params
+          .order()
+          .map((o) => o.join('.'))
+          .join(',')
+      : undefined,
+    _limit: params.limit().toString(),
+    _search: params.search() || undefined,
+    _count: params.includeCount()
+      ? params.includeCount().toString()
```
Comment (dcsaszar) [inline src/Data/toClientParams.ts:26]: * [ ] FYI `params.includeCount().toString()` is a very elaborate way to say `true`

File: src/Data/toClientParams.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+type IndexCallback = NonNullable<DataConnection['index']>
+type IndexParams = Parameters<IndexCallback>[0]
+
+interface OperatorSpec {
+  operator: 'equals' | 'notEquals'
+  opCode: 'eq' | 'neq'
+  value: string
+}
+
+export function toClientParams(params: IndexParams) {
```
Comment (dcsaszar) [inline src/Data/toClientParams.ts:13]: * [ ] Missing a return type.

File: src/Data/toClientParams.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    ...toClientFilterParam(params.filters()),
+    _continuation: params.continuation(),
+    _order: params.order().length
+      ? params
+          .order()
+          .map((o) => o.join('.'))
+          .join(',')
+      : undefined,
+    _limit: params.limit().toString(),
+    _search: params.search() || undefined,
```
Comment (dcsaszar) [inline src/Data/toClientParams.ts:24]: * [ ] FYI Not sure about the `|| undefined`. Looks like a safeguard against accidental empty-string searches, but don't we hide potential application errors with this?

File: src/Data/toClientParams.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      ? params.includeCount().toString()
+      : undefined,
+  }
+}
+
+function toClientFilterParam(filters: Record<string, OperatorSpec>) {
+  const params: Record<string, string> = {}
+
+  Object.keys(filters).forEach((name) => {
+    const { opCode, value } = filters[name] as OperatorSpec
```
Comment (dcsaszar) [inline src/Data/toClientParams.ts:35]: * [ ] FYI We can get rid of casting thanks to `Object.entries`

```ts
Object.entries(filters).forEach(([name, { opCode, value }]) => {
```

File: src/utils/convertBlobAttributes.ts [hunk 6]
```diff
@@ -23,20 +23,23 @@ async function blobToBinary(blob: Blob | File): Promise<{
   dataBase64: string
   filename: string
 
-  // TODO: Remove for pisa
-  contentLength: number
-  contentType: string
-  _id: string
+  contentLength?: number // only needed for "localStorage" fake
+  contentType?: string // only needed for "localStorage" fake
+  _id?: string // only needed for "localStorage" fake
```
Comment (dcsaszar) [inline src/utils/convertBlobAttributes.ts:28]: * [ ] FYI Currently a single headline comment would do.
* [ ] FYI I'd not call it "fake" b/c there are valid use cases for local storage.

File: src/utils/pisaDataBinaryToUrl.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import { pisaClient, pisaUrl } from '../Data/pisaClient'
+import { FullDataBinary } from './dataBinaryToUrl'
+
+export async function pisaDataBinaryToUrl(binary: FullDataBinary) {
```
Comment (dcsaszar) [inline src/utils/pisaDataBinaryToUrl.ts:4]: * [ ] Missing return type. However, as long as dataBinaryToUrl is the only importing module, we could move it there as a private helper.

File: vite.config.ts [hunk 8]
```diff
@@ -16,7 +16,7 @@ export default defineConfig(({ mode }) => {
 
   const PISA_URL = env.PISA_URL
   const enablePisa = !!PISA_URL
-  if (enablePisa) console.log('Pisa included in the build.')
+  if (enablePisa) console.log('Pisa is enabled.')
```
Comment (dcsaszar) [inline vite.config.ts:19]: * [ ] FYI I realized that in practice this message is noisy. Personally I'd like to get rid of it.

**Review by dcsaszar — APPROVED** [review]
Nice. Approved b/c mandatory feedback is about specifying return types, which could also be handled in a follow-up.

File: src/Data/toClientParams.ts (hunk 3 above)
Comment (apepper) [reply src/Data/toClientParams.ts:13]: Why is this needed? The derived types look fine to me.

File: src/Data/toClientParams.ts (hunk 4 above)
Comment (apepper) [reply src/Data/toClientParams.ts:24]: This code is inspired by the SDK.

File: src/utils/pisaDataBinaryToUrl.ts (hunk 7 above)
Comment (apepper) [reply src/utils/pisaDataBinaryToUrl.ts:4]: Same as https://github.com/Scrivito/scrivito-portal-app/pull/381#discussion_r1578102346,

File: vite.config.ts (hunk 8 above)
Comment (apepper) [reply vite.config.ts:19]: Let's see, if I remove it eventually.

File: src/Data/toClientParams.ts (hunk 4 above)
Comment (dcsaszar) [reply src/Data/toClientParams.ts:24]: > This code is inspired by the SDK.

I don't understand. If the SDK would autoconvert `''` to `undefined` we wouldn't need this the fallback here at all.

Comment (apepper) [reply src/Data/toClientParams.ts:24]: It's inspired by `js/src/scrivito_sdk/data_integration/create_standard_api_connection.ts#L109`. This is actually the behavior of `provideDataClass` with `restApi`.

**Pull request merged**